diff --git a/app/Resources/views/facebook/registration.html.twig b/app/Resources/views/facebook/registration.html.twig
new file mode 100644
index 0000000..af752a5
--- /dev/null
+++ b/app/Resources/views/facebook/registration.html.twig
@@ -0,0 +1,9 @@
+{% extends 'base.html.twig' %}
+
+{% block body %}
+    {{ form_start(form) }}
+        {{ form_widget(form) }}
+
+        <button type="submit">Register!</button>
+    {{ form_end(form) }}
+{% endblock %}
diff --git a/src/AppBundle/Controller/FacebookConnectController.php b/src/AppBundle/Controller/FacebookConnectController.php
index ee56036..cf76088 100644
--- a/src/AppBundle/Controller/FacebookConnectController.php
+++ b/src/AppBundle/Controller/FacebookConnectController.php
@@ -2,6 +2,9 @@
 
 namespace AppBundle\Controller;
 
+use AppBundle\Entity\User;
+use AppBundle\Form\FacebookRegistrationType;
+use League\OAuth2\Client\Provider\FacebookUser;
 use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
 use Symfony\Bundle\FrameworkBundle\Controller\Controller;
 use Symfony\Component\HttpFoundation\Request;
@@ -33,4 +36,51 @@ class FacebookConnectController extends Controller
     {
         // will not be reached!
     }
+
+    /**
+     * @Route("/connect/facebook/registration", name="connect_facebook_registration")
+     */
+    public function finishRegistration(Request $request)
+    {
+        // stored in the session in FacebookAuthenticator
+        /** @var FacebookUser $facebookUser */
+        $facebookUser = $request->getSession()->get('facebook_user');
+        if (!$facebookUser) {
+            throw $this->createNotFoundException('How did you get here without user information!?');
+        }
+        $user = new User();
+        $user->setFacebookId($facebookUser->getId());
+        $user->setEmail($facebookUser->getEmail());
+
+        $form = $this->createForm(new FacebookRegistrationType(), $user);
+
+        $form->handleRequest($request);
+        if ($form->isValid()) {
+            // encode the password manually
+            $plainPassword = $form['plainPassword']->getData();
+            $encodedPassword = $this->get('security.password_encoder')
+                ->encodePassword($user, $plainPassword);
+            $user->setPassword($encodedPassword);
+
+            $em = $this->getDoctrine()->getManager();
+            $em->persist($user);
+            $em->flush();
+
+            // remove the session information
+            $request->getSession()->remove('facebook_user');
+
+            // log the user in manually
+            $guardHandler = $this->container->get('security.authentication.guard_handler');
+            return $guardHandler->authenticateUserAndHandleSuccess(
+                $user,
+                $request,
+                $this->container->get('app.facebook_authenticator'),
+                'main' // the firewall key
+            );
+        }
+
+        return $this->render('facebook/registration.html.twig', array(
+            'form' => $form->createView()
+        ));
+    }
 }
diff --git a/src/AppBundle/Form/FacebookRegistrationType.php b/src/AppBundle/Form/FacebookRegistrationType.php
new file mode 100644
index 0000000..b3da268
--- /dev/null
+++ b/src/AppBundle/Form/FacebookRegistrationType.php
@@ -0,0 +1,37 @@
+<?php
+
+namespace AppBundle\Form;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class FacebookRegistrationType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('username', 'text')
+            ->add('email', 'text', array(
+                'disabled' => true
+            ))
+            ->add('plainPassword', 'repeated', array(
+                'mapped' => false, // allows this to not be a real property on User
+                'type' => 'password',
+                'first_options'  => array('label' => 'Password'),
+                'second_options'  => array('label' => 'Password again'),
+            ));
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults(array(
+            'data_class' => 'AppBundle\Entity\User'
+        ));
+    }
+
+    public function getName()
+    {
+        return 'app_bundle_user_registration_type';
+    }
+}
diff --git a/src/AppBundle/OAuth/FinishRegistrationException.php b/src/AppBundle/OAuth/FinishRegistrationException.php
new file mode 100644
index 0000000..a806256
--- /dev/null
+++ b/src/AppBundle/OAuth/FinishRegistrationException.php
@@ -0,0 +1,41 @@
+<?php
+
+namespace AppBundle\OAuth;
+
+use Exception;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+
+class FinishRegistrationException extends AuthenticationException
+{
+    private $userInformation;
+
+    /**
+     * @param mixed         $userInfo Any info to be used to help registration
+     * @param string        $message
+     * @param integer       $code
+     * @param \Exception    $previous
+     */
+    public function __construct($userInfo, $message = "", $code = 0, Exception $previous = null)
+    {
+        $this->userInformation = $userInfo;
+
+        parent::__construct($message, $code, $previous);
+    }
+
+    public function getUserInformation()
+    {
+        return $this->userInformation;
+    }
+
+    /**
+     * This method should *not* be called normally. The purpose
+     * of this exception is to be a "flag" for code that will
+     * redirect the user to some "finish registration" page. If
+     * this message is shown to your user, then you're missing
+     * this piece of your code.
+     */
+    public function getMessageKey()
+    {
+        return 'You need to finish registration to login.';
+    }
+}
diff --git a/src/AppBundle/Security/FacebookAuthenticator.php b/src/AppBundle/Security/FacebookAuthenticator.php
index 3472d80..aff9dc9 100644
--- a/src/AppBundle/Security/FacebookAuthenticator.php
+++ b/src/AppBundle/Security/FacebookAuthenticator.php
@@ -3,18 +3,18 @@
 namespace AppBundle\Security;
 
 use AppBundle\Entity\User;
+use AppBundle\OAuth\FinishRegistrationException;
+use AppBundle\Security\Helper\FinishRegistrationBehavior;
 use AppBundle\Security\Helper\SaveAuthFailureMessage;
-use AppBundle\Security\Helper\RedirectToLoginWhenAnonymous;
 use AppBundle\Security\Helper\PreviousUrlHelper;
 use Doctrine\ORM\EntityManager;
 use League\OAuth2\Client\Provider\ResourceOwnerInterface;
 use League\OAuth2\Client\Token\AccessToken;
 use League\OAuth2\Client\Provider\Facebook;
 use League\OAuth2\Client\Provider\FacebookUser;
-use Symfony\Component\HttpFoundation\RedirectResponse;
-use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\Routing\RouterInterface;
 use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RedirectResponse;
 use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
 use Symfony\Component\Security\Core\Exception\AuthenticationException;
 
@@ -22,6 +22,7 @@ class FacebookAuthenticator extends AbstractSocialAuthenticator
 {
     use PreviousUrlHelper;
     use SaveAuthFailureMessage;
+    use FinishRegistrationBehavior;
 
     private $facebook;
     private $em;
@@ -61,15 +62,10 @@ class FacebookAuthenticator extends AbstractSocialAuthenticator
         $user = $this->em->getRepository('AppBundle:User')
                     ->findOneBy(array('email' => $email));
 
-        // 3) no user? Perhaps you just want to create one
-        //      or maybe you want to redirect to a registration (in that case, keep reading_
+        // 3) no user? Redirect to finish registration
         if (!$user) {
-            $user = new User();
-            $user->setUsername($email);
-            $user->setEmail($email);
-            // set an un-encoded password, which basically makes it *not* possible
-            // to login with any password
-            $user->setPassword('no password');
+            // throw a special exception we created - see onAuthenticaitonFailure
+            throw new FinishRegistrationException($facebookUser);
         }
 
         // make sure the Facebook user is set
@@ -96,6 +92,12 @@ class FacebookAuthenticator extends AbstractSocialAuthenticator
 
     public function onAuthenticationFailure(Request $request, AuthenticationException $exception)
     {
+        if ($this->requiresFinishRegistration($exception)) {
+            $registrationUrl = $this->router->generate('connect_facebook_registration');
+
+            return $this->sendToFinishRegistration($registrationUrl, $request, $exception);
+        }
+
         $this->saveAuthenticationErrorToSession($request, $exception);
 
         $loginUrl = $this->router->generate('security_login');
diff --git a/src/AppBundle/Security/Helper/FinishRegistrationBehavior.php b/src/AppBundle/Security/Helper/FinishRegistrationBehavior.php
new file mode 100644
index 0000000..3d233c1
--- /dev/null
+++ b/src/AppBundle/Security/Helper/FinishRegistrationBehavior.php
@@ -0,0 +1,42 @@
+<?php
+
+namespace AppBundle\Security\Helper;
+
+use AppBundle\OAuth\FinishRegistrationException;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+
+/**
+ * Use this trait if sometimes your authenticator requires people
+ * to "finish registration" before logging in.
+ */
+trait FinishRegistrationBehavior
+{
+    /**
+     * Should we send the user to finish registration?
+     *
+     * @param AuthenticationException $e
+     * @return bool
+     */
+    protected function requiresFinishRegistration(AuthenticationException $e)
+    {
+        return $e instanceof FinishRegistrationException;
+    }
+
+    /**
+     * Call this from within your onAuthenticationFailure() method.
+     *
+     * @param string $registrationUrl
+     * @param Request $request
+     * @param FinishRegistrationException $e
+     * @return RedirectResponse
+     */
+    protected function sendToFinishRegistration($registrationUrl, Request $request, FinishRegistrationException $e)
+    {
+        // save the user information!
+        $request->getSession()->set('guard.finish_registration.user_information', $e->getUserInformation());
+
+        return new RedirectResponse($registrationUrl);
+    }
+}
