diff --git a/app/Resources/views/security/login.html.twig b/app/Resources/views/security/login.html.twig
index 0d5b3a4..23720ac 100644
--- a/app/Resources/views/security/login.html.twig
+++ b/app/Resources/views/security/login.html.twig
@@ -20,5 +20,7 @@
 
         <br/>
         <button type="submit">Login</button>
+
+        <a href="{{ path('connect_facebook') }}">Login with Facebook</a>
     </form>
 {% endblock %}
diff --git a/app/config/parameters.yml b/app/config/parameters.yml
index f5e0347..b579b4a 100644
--- a/app/config/parameters.yml
+++ b/app/config/parameters.yml
@@ -10,3 +10,6 @@ parameters:
     mailer_user: null
     mailer_password: null
     secret: a3250d79a622c2f1d034943d9ff4397d500552dd
+    facebook_app_id:  XXXX
+    facebook_app_secret: XXXX
+
diff --git a/app/config/parameters.yml.dist b/app/config/parameters.yml.dist
index 886c92b..2833651 100644
--- a/app/config/parameters.yml.dist
+++ b/app/config/parameters.yml.dist
@@ -17,3 +17,5 @@ parameters:
 
     # A secret key that's used to generate certain security-related tokens
     secret:            ThisTokenIsNotSoSecretChangeIt
+    facebook_app_id:  XXXX
+    facebook_app_secret: XXXX
diff --git a/app/config/security.yml b/app/config/security.yml
index 23ce35a..d802b03 100644
--- a/app/config/security.yml
+++ b/app/config/security.yml
@@ -22,5 +22,6 @@ security:
                 authenticators:
                     - app.form_login_authenticator
                     - app.api_token_authenticator
+                    - app.facebook_authenticator
                 # by default, use the start() function from FormLoginAuthenticator
                 entry_point: app.form_login_authenticator
diff --git a/app/config/services.yml b/app/config/services.yml
index 5dadd04..843345e 100644
--- a/app/config/services.yml
+++ b/app/config/services.yml
@@ -11,3 +11,19 @@ services:
     app.api_token_authenticator:
         class: AppBundle\Security\ApiTokenAuthenticator
         arguments: ["@doctrine.orm.entity_manager"]
+
+    app.facebook_provider:
+        class: League\OAuth2\Client\Provider\Facebook
+        arguments:
+            -
+                clientId: %facebook_app_id%
+                clientSecret: %facebook_app_secret%
+                graphApiVersion: v2.3
+                redirectUri: "@=service('router').generate('connect_facebook_check', {}, true)"
+
+    app.facebook_authenticator:
+        class: AppBundle\Security\FacebookAuthenticator
+        arguments:
+            # you can also inject the services you need individually
+            # except for the router, as it causes a circular reference problem
+            - @service_container
diff --git a/src/AppBundle/Controller/FacebookConnectController.php b/src/AppBundle/Controller/FacebookConnectController.php
new file mode 100644
index 0000000..cfc976e
--- /dev/null
+++ b/src/AppBundle/Controller/FacebookConnectController.php
@@ -0,0 +1,29 @@
+<?php
+
+namespace AppBundle\Controller;
+
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+class FacebookConnectController extends Controller
+{
+    /**
+     * @Route("/connect/facebook", name="connect_facebook")
+     */
+    public function connectFacebookAction(Request $request)
+    {
+        // redirects to Facebook - see FacebookAuthenticator::start()
+        return $this->get('app.facebook_authenticator')->start($request);
+    }
+
+    /**
+     * @Route("/connect/facebook-check", name="connect_facebook_check")
+     */
+    public function connectFacebookActionCheck()
+    {
+        // will not be reached!
+    }
+}
diff --git a/src/AppBundle/Entity/User.php b/src/AppBundle/Entity/User.php
index baeeb95..6421105 100644
--- a/src/AppBundle/Entity/User.php
+++ b/src/AppBundle/Entity/User.php
@@ -51,6 +51,11 @@ class User implements UserInterface
      */
     private $apiToken;
 
+    /**
+     * @ORM\Column(type="string", length=50, nullable=true)
+     */
+    private $facebookId;
+
     public function __construct()
     {
         $this->apiToken = base_convert(sha1(uniqid(mt_rand(), true)), 16, 36);
@@ -136,4 +141,14 @@ class User implements UserInterface
     {
         $this->apiToken = $apiToken;
     }
+
+    public function getFacebookId()
+    {
+        return $this->facebookId;
+    }
+
+    public function setFacebookId($facebookId)
+    {
+        $this->facebookId = $facebookId;
+    }
 }
diff --git a/src/AppBundle/Security/FacebookAuthenticator.php b/src/AppBundle/Security/FacebookAuthenticator.php
index f0a0435..79d62f4 100644
--- a/src/AppBundle/Security/FacebookAuthenticator.php
+++ b/src/AppBundle/Security/FacebookAuthenticator.php
@@ -2,6 +2,17 @@
 
 namespace AppBundle\Security;
 
+use AppBundle\Entity\User;
+use Doctrine\ORM\EntityManager;
+use KnpU\Guard\Exception\CustomAuthenticationException;
+use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
+use League\OAuth2\Client\Provider\Facebook;
+use League\OAuth2\Client\Provider\FacebookUser;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Security;
 use KnpU\Guard\AbstractGuardAuthenticator;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
@@ -11,38 +22,137 @@ use Symfony\Component\Security\Core\User\UserProviderInterface;
 
 class FacebookAuthenticator extends AbstractGuardAuthenticator
 {
+    private $container;
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
     public function getCredentials(Request $request)
     {
-        // todo
+        if ($request->getPathInfo() != '/connect/facebook-check') {
+            // skip authentication unless we're on this URL!
+            return null;
+        }
+
+        if ($code = $request->query->get('code')) {
+            return $code;
+        }
+
+        // no code! Something went wrong. Quite probably the user denied our app access
+        // you could read the error, error_code, error_description, error_reason query params
+        // http://localhost:8000/connect/facebook-check?error=access_denied&error_code=200&error_description=Permissions+error&error_reason=user_denied&state=S2fKgHJSZSJM0Qs2fhKL6USZP50KSBHc#_=_
+        throw CustomAuthenticationException::createWithSafeMessage(
+            'There was an error getting access from Facebook. Please try again.'
+        );
     }
 
     public function getUser($authorizationCode, UserProviderInterface $userProvider)
     {
-        // todo
+        try {
+            // the credentials are really the access token
+            $accessToken = $this->getFacebookOAuthProvider()->getAccessToken(
+                'authorization_code',
+                ['code' => $authorizationCode]
+            );
+        } catch (IdentityProviderException $e) {
+            // probably the authorization code has been used already
+            $response = $e->getResponseBody();
+            $errorCode = $response['error']['code'];
+            $message = $response['error']['message'];
+
+            throw CustomAuthenticationException::createWithSafeMessage(
+                'There was an error logging you into Facebook - code '.$errorCode
+            );
+        }
+
+        /** @var FacebookUser $facebookUser */
+        $facebookUser = $this->getFacebookOAuthProvider()->getResourceOwner($accessToken);
+        $email = $facebookUser->getEmail();
+
+        $em = $this->container->get('doctrine')->getManager();
+
+        // 1) have they logged in with Facebook before? Easy!
+        $existingUser = $em->getRepository('AppBundle:User')
+            ->findOneBy(array('facebookId' => $facebookUser->getId()));
+        if ($existingUser) {
+            return $existingUser;
+        }
+
+        // 2) do we have a matching user by email?
+        $user = $em->getRepository('AppBundle:User')
+                    ->findOneBy(array('email' => $email));
+
+        // 3) no user? Perhaps you just want to create one
+        //      or maybe you want to redirect to a registration (in that case, keep reading_
+        if (!$user) {
+            $user = new User();
+            $user->setUsername($email);
+            $user->setEmail($email);
+            // set an un-encoded password, which basically makes it *not* possible
+            // to login with any password
+            $user->setPassword('no password');
+        }
+
+        // make sure the Facebook user is set
+        $user->setFacebookId($facebookUser->getId());
+        $em->persist($user);
+        $em->flush();
+
+        return $user;
     }
 
     public function checkCredentials($credentials, UserInterface $user)
     {
-        // todo
+        // do nothing - the fact that the access token worked means that
+        // our app has been authorized with Facebook
     }
 
     public function onAuthenticationFailure(Request $request, AuthenticationException $exception)
     {
-        // todo
+        // this would happen if something went wrong in the OAuth flow
+        $request->getSession()->set(Security::AUTHENTICATION_ERROR, $exception);
+
+        $url = $this->container->get('router')
+            ->generate('security_login');
+
+        return new RedirectResponse($url);
     }
 
     public function onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)
     {
-        // todo
+        // todo - remove needing this crazy thing
+        $targetPath = $request->getSession()->get('_security.'.$providerKey.'.target_path');
+
+        if (!$targetPath) {
+            $router = $this->container->get('router');
+            $targetPath = $router->generate('homepage');
+        }
+
+        return new RedirectResponse($targetPath);
     }
 
     public function supportsRememberMe()
     {
-        // todo
+        return true;
     }
 
     public function start(Request $request, AuthenticationException $authException = null)
     {
-        // todo
+        $authUrl = $this->getFacebookOAuthProvider()->getAuthorizationUrl([
+            // these are actually the default scopes
+            'scopes' => ['public_profile', 'email'],
+        ]);
+
+        return new RedirectResponse($authUrl);
+    }
+
+    /**
+     * @return Facebook
+     */
+    private function getFacebookOAuthProvider()
+    {
+        return $this->container->get('app.facebook_provider');
     }
 }
