diff --git a/src/AppBundle/Security/FormLoginAuthenticator.php b/src/AppBundle/Security/FormLoginAuthenticator.php
index 6623768..efe6487 100644
--- a/src/AppBundle/Security/FormLoginAuthenticator.php
+++ b/src/AppBundle/Security/FormLoginAuthenticator.php
@@ -4,9 +4,11 @@ namespace AppBundle\Security;
 
 use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
 use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
 use Symfony\Component\Security\Guard\Authenticator\AbstractFormLoginAuthenticator;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
 use Symfony\Component\Security\Core\Exception\BadCredentialsException;
 use Symfony\Component\Security\Core\Security;
 use Symfony\Component\Security\Core\User\UserInterface;
@@ -67,6 +69,21 @@ class FormLoginAuthenticator extends AbstractFormLoginAuthenticator
         return true;
     }
 
+    public function onAuthenticationFailure(Request $request, AuthenticationException $exception)
+    {
+        // AJAX! Maybe return some JSON
+        if ($request->isXmlHttpRequest()) {
+            return new JsonResponse(
+                // you could translate the message
+                array('message' => $exception->getMessageKey()),
+                403
+            );
+        }
+
+        // for non-AJAX requests, return the normal redirect
+        return parent::onAuthenticationFailure($request, $exception);
+    }
+
     protected function getLoginUrl()
     {
         return $this->container->get('router')
