diff --git a/src/AppBundle/Security/FormLoginAuthenticator.php b/src/AppBundle/Security/FormLoginAuthenticator.php
index 91e167f..95231b3 100644
--- a/src/AppBundle/Security/FormLoginAuthenticator.php
+++ b/src/AppBundle/Security/FormLoginAuthenticator.php
@@ -5,6 +5,9 @@ namespace AppBundle\Security;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 use Symfony\Component\Security\Guard\Authenticator\AbstractFormLoginAuthenticator;
 use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\Security;
 use Symfony\Component\Security\Core\User\UserInterface;
 use Symfony\Component\Security\Core\User\UserProviderInterface;
 
@@ -19,26 +22,47 @@ class FormLoginAuthenticator extends AbstractFormLoginAuthenticator
 
     public function getCredentials(Request $request)
     {
-        // TODO: Implement getCredentials() method.
+        if ($request->getPathInfo() != '/login_check') {
+            return;
+        }
+
+        $username = $request->request->get('_username');
+        $request->getSession()->set(Security::LAST_USERNAME, $username);
+        $password = $request->request->get('_password');
+
+        return array(
+            'username' => $username,
+            'password' => $password
+        );
     }
 
     public function getUser($credentials, UserProviderInterface $userProvider)
     {
-        // TODO: Implement getUser() method.
+        $username = $credentials['username'];
+
+        return $userProvider->loadUserByUsername($username);
     }
 
     public function checkCredentials($credentials, UserInterface $user)
     {
-        // TODO: Implement checkCredentials() method.
+        $plainPassword = $credentials['password'];
+        $encoder = $this->container->get('security.password_encoder');
+        if (!$encoder->isPasswordValid($user, $plainPassword)) {
+            return false;
+        }
+
+        return true;
     }
 
     protected function getLoginUrl()
     {
-        // TODO: Implement getLoginUrl() method.
+        return $this->container->get('router')
+            ->generate('security_login');
     }
 
     protected function getDefaultSuccessRedirectUrl()
     {
-        // TODO: Implement getDefaultSuccessRedirectUrl() method.
+        return $this->container->get('router')
+            ->generate('homepage');
     }
 }
