diff --git a/src/AppBundle/Security/FormLoginAuthenticator.php b/src/AppBundle/Security/FormLoginAuthenticator.php
index c8954a9..8461468 100644
--- a/src/AppBundle/Security/FormLoginAuthenticator.php
+++ b/src/AppBundle/Security/FormLoginAuthenticator.php
@@ -4,34 +4,64 @@ namespace AppBundle\Security;
 
 use KnpU\Guard\Authenticator\AbstractFormLoginAuthenticator;
 use Symfony\Component\HttpFoundation\Request;
-use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\Security;
 use Symfony\Component\Security\Core\User\UserInterface;
 use Symfony\Component\Security\Core\User\UserProviderInterface;
 
 class FormLoginAuthenticator extends AbstractFormLoginAuthenticator
 {
+    private $router;
+
+    private $encoder;
+
+    public function __construct(RouterInterface $router, UserPasswordEncoderInterface $encoder)
+    {
+        $this->router = $router;
+        $this->encoder = $encoder;
+    }
+
     public function getCredentials(Request $request)
     {
-        // TODO: Implement getCredentials() method.
+        if ($request->getPathInfo() != '/login_check' || !$request->isMethod('POST')) {
+            return;
+        }
+
+        $username = $request->request->get('_username');
+        $request->getSession()->set(Security::LAST_USERNAME, $username);
+        $password = $request->request->get('_password');
+
+        return array(
+            'username' => $username,
+            'password' => $password
+        );
     }
 
     public function getUser($credentials, UserProviderInterface $userProvider)
     {
-        // TODO: Implement getUser() method.
+        $username = $credentials['username'];
+
+        return $userProvider->loadUserByUsername($username);
     }
 
     public function checkCredentials($credentials, UserInterface $user)
     {
-        // TODO: Implement checkCredentials() method.
+        $plainPassword = $credentials['password'];
+        if (!$this->encoder->isPasswordValid($user, $plainPassword)) {
+            // throw any AuthenticationException
+            throw new BadCredentialsException();
+        }
     }
 
     protected function getLoginUrl()
     {
-        // TODO: Implement getLoginUrl() method.
+        return $this->router->generate('security_login');
     }
 
     protected function getDefaultSuccessRedirectUrl()
     {
-        // TODO: Implement getDefaultSuccessRedirectUrl() method.
+        return $this->router->generate('homepage');
     }
 }
